### PR TITLE
Append to existing trailers in generated squash commit message

### PR DESCRIPTION
**Pull request by jpraet**
If the PR description already contains a trailer section (for example with a Signed-off-by: header), then just append the gitea-generated trailers (Co-authered-by, Reviewed-on, ...) to this existing section.

### CODE REVIEW

File: services/pull/pull.go [hunk 1]
```diff
@@ -573,7 +574,10 @@ func GetSquashMergeCommitMessages(pr *models.PullRequest) string {
 	stringBuilder.WriteString(pr.Issue.Content)
 	if stringBuilder.Len() > 0 {
 		stringBuilder.WriteRune('\n')
-		stringBuilder.WriteRune('\n')
+		commitMessageTrailersPattern := regexp.MustCompile(`(^|.*\n\n)([\w-]+: [^\n]+)(\n[\w-]+: [^\n]+)*$`)
```
Comment (zeripath) [inline services/pull/pull.go]: Compile regexps once. Recompiling them every request is a bad idea.

**Review by zeripath — CHANGES_REQUESTED** [review]
as per comment

**Comment by zeripath** [comment]
Thanks for the change. I'm still a bit suspicious about the regexp.

For example:

```
(^l.*\n\n)
```

Why not:

```
(?:^|\n\n)
```

Are you actually capturing the groups for some reason? If not use non capturing. (It can actually make a significant difference in time and memory allocation.)

Don't use `.*` if you don't need it. In fact in non 's' mode`.` doesn't capture `\n` so I'm not certain what the `.*` is looking for.

Your regexp also seems to require that trailers don't end with a newline which seems contrary to git which prefers messages end with a new line.

**Comment by jpraet** [comment]
I'm a bit rusty with regex. It should be better now: https://regex101.com/r/UDBQe0/1

File: services/pull/pull.go [hunk 2]
```diff
@@ -516,6 +517,8 @@ func CloseRepoBranchesPulls(doer *models.User, repo *models.Repository) error {
 	return nil
 }
 
+var commitMessageTrailersPattern = regexp.MustCompile(`(?:^|\n\n)(?:[\w-]+: [^\n]+\n*)+$`)
```
Comment (zeripath) [inline services/pull/pull.go]: OK testing from git interpret-trailers I think this needs to be, at least:

```suggestion
var commitMessageTrailersPattern = regexp.MustCompile(`(?:^|\n\n)(?:[\w-]+[:=][^\n]+\n)*(?:[\w-]+: [^\n]+)\n*$`)
```

However trailers can be folded, see https://git-scm.com/docs/git-interpret-trailers

> When reading trailers, there can be whitespaces after the token, the separator and the value. There can also be whitespaces inside the token and the value. The value may be split over multiple lines with each subsequent line starting with whitespace, like the "folding" in RFC 822.

Which would make this:

```suggestion
var commitMessageTrailersPattern = regexp.MustCompile(`(?:^|\n\n)(?:[\w-]+[:=][^\n]+\n(?: [^\n]*\n)*)*(?:[\w-]+: [^\n]+(?:\n [^\n]*)*)\n*$`)
```

(I think)

Comment (zeripath) [reply services/pull/pull.go]:
```
echo -n 'Fix #1546                                                                         [19:26:27]

Signed-off-by: Not a trailer

Signed-off-by:Alice <alice@example.com>
Signed-off-by:   Bob <bob@example.com>
' | git interpret-trailers --parse
```

```
echo -n 'Fix #1546                                                                         [19:26:27]

Signed-off-by:Not a trailer

Signed-off-by:Alice <alice@example.com>
Signed-off-by:   Bob <bob@example.com>' | git interpret-trailers --parse
```

```
echo -n 'Fix #1546

Signed-off-by:Not a trailer

Signed-off-by:Alice <alice@example.com>
Signed-off-by:   Bob <bob@example.com>


' | git interpret-trailers --parse
```

```
echo -n 'Fix #1546

Signed-off-by:Not a trailer

Signed-off-by:Alice <alice@example.com>
Folded: This is a folded value
 with folded results
   preceding spaces?
Signed-off-by:   Bob <bob@example.com>' | git interpret-trailers --parse
```

**Review by zeripath — APPROVED** [review]

File: services/pull/pull_test.go [hunk 3]
```diff
@@ -5,4 +5,27 @@
...
+)
+
 // TODO TestPullRequest_PushToBaseRepo
+
+func TestPullRequest_CommitMessageTrailersPattern(t *testing.T) {
+	// Not a valid trailer section
+	assert.False(t, commitMessageTrailersPattern.MatchString(""))
+	assert.False(t, commitMessageTrailersPattern.MatchString("No trailer."))
+	assert.False(t, commitMessageTrailersPattern.MatchString("Signed-off-by: Bob <bob@example.com>\nNot a trailer due to following text."))
+	assert.False(t, commitMessageTrailersPattern.MatchString("Message body not correctly separated from trailer section by empty line.\nSigned-of-by: Bob <bob@example.com>"))
```
Comment (noerw) [inline services/pull/pull_test.go]: tiny typo
```suggestion
	assert.False(t, commitMessageTrailersPattern.MatchString("Message body not correctly separated from trailer section by empty line.\nSigned-off-by: Bob <bob@example.com>"))
```

**Review by noerw — APPROVED** [review]

**Pull request merged**